# TIL - 2023.11.01(수)
## 자바 비동기

### CompletableFuture
**자바 비동기 처리하기**
```java
CompletableFuture.runAsync(() -> log.info("hello"));
```
- 자바에서 제공하는 `CompletableFuture`를 사용하면 비동기 메서드를 처리할 수 있다

```
[ForkJoinPool.commonPool-worker-1] CompletableFutureTest -- run async
```
- 자바에서는 `CompletableFuture`를 `ForkJoinPool`을 사용하여 별도의 스레드에서 처리해준다

```java
// 비동기 작업 수행하기
Future<String> result = CompletableFuture.supplyAsync(() -> "result");
        
// 비동기 수행 결과 가져오기
result.get();
```
- `CompletableFuture`로 비동기 메서드를 수행하면 `Future` 타입을 반환한다
- `Future.get`을 호출하여 비동기 메서드의 수행 결과를 가져올 수 있다. 해당 메서드는 블로킹 메소드이기 때문에, 값을 반환할 때까지 대기해야 한다.

**자바 비동기 체이닝**

만약 비동기 메서드 수행이 완료된 후 추가적인 처리를 하고 싶다면 어떻게 해야할까?
- `CompletableFuture`는 여러 가지 체이닝 메서드들을 지원하기 때문에 간편하게 후처리가 가능하다

```java
CompletableFuture
    .runAsync(() -> log.info("run async"))
    // 동일한 스레드에서 수행
    .thenRun(() -> log.info("then run async"));

[ForkJoinPool.commonPool-worker-1] CompletableFutureTest -- run async
[ForkJoinPool.commonPool-worker-1] CompletableFutureTest -- then run async
```
- 예를 들어, `runAsync`의 수행 결과에 `thenRun`을 체이닝하여 비동기 수행 후 새로운 작업을 순차적으로 실행할 수 있다
- `thenRun`는 이전에 사용했던 스레드를 동이랗게 사용하여 비동기 메서드를 처리해준다

```java
CompletableFuture
    .runAsync(() -> log.info("run async"))
    // 동일한 스레드에서 수행
    .thenRunAsync(() -> log.info("then run async"));

[ForkJoinPool.commonPool-worker-1] CompletableFutureTest -- runAsync
[ForkJoinPool.commonPool-worker-2] CompletableFutureTest -- thenRunAsync
```
- `thenRunAsync`를 사용하면 다른 스레드를 사용하여 비동기 메서드 처리도 가능하다

```java
CompletableFuture
    .supplyAsync(() -> "hello")
    .thenApply(res -> res + " bye")
    .thenAccept(res -> log.info("thenAccept, {}", res));
```
- 그외에도 다양한 체이닝 메서드들을 지원한다
- `thenApply`는 `Function` 타입을 인자로 받으므로, 이전 단계의 결과를 받아 새롭게 변형하여 반환할 수 있다
- `thenAccept`는 `Consumer` 타입을 인자로 받으므로, 이전 단계의 결과를 받아와서 처리가 가능하다

**자바 비동기 예외처리**
```java
CompletableFuture
    .supplyAsync(() -> throw new RuntimeException());
```
- 만약 위와 같이 비동기 체이닝 메서드 수행 중, 예외가 발생한다면 어떻게 해야할까?

```java
Future<String> cf = CompletableFuture
                        .supplyAsync(() -> throw new RuntimeException("async error occurred!")
                        .thenApplyAsync(res -> res + " hello")
                        .exceptionally(e -> return "error")
```
- `CompletableFuture` 수행 중에 발생한 예외는 `exceptionally`를 사용해 처리할 수 있다
- 만약 `supplyAsync` 메서드에서 예외가 발생한다면, 뒤에 등록된 체이닝 메서드들은 수행되지 않고 바로 `exceptionally`가 수행된다
- `exceptionally` 메서드에서 "error"를 반환하므로, `cf.get()`을 실행하면 "error" 를 반환할 것이다

```java
CompletableFuture
            .supplyAsync(() -> throw new RuntimeException("async error occurred!"))
            .thenApplyAsync(res -> res + " hello")
            .handle((s, ex) -> ex != null ? "error" : s);
```
- 혹은 `handle`을 사용하여 예외 처리가 가능하다
- `handle`은 `finally` 처럼 예외 발생 여부와 관계 없이 항상 수행된다
- `handle`은 `BiFunction` 타입을 인자로 받는데, `BiFunction`은 인자로 이전 스테이지에서 넘겨준 값과 예외 값을 전달받는다
- 만약 예외가 발생하면 `s=null`이 되고 `ex`에는 예외 정보가 들어가 있을 것이다. 이를 활용하여 비동기 예외 처리가 가능하다

**자바 비동기 중첩하기**
```java
// Future<Future<String>> 타입
CompletableFuture
    .supplyAsnyc(() -> CompletableFuture.supplyAsync(() -> "hello"));
```
- 만약 위와 같이 `CompletableFuture`가 중첩된다면 어떻게 처리해야 할까?

```java
// Future<String>
CompletableFuture
        .supplyAsync(() -> "hello")
        .thenCompose(res -> 
            CompletableFuture
                .supplyAsync(() -> return res + " compose")
        );
```
- 이럴 때는 `CompletableFuture.thenCompose`를 사용하면 된다
- 마치 `Stream.flatMap` 처럼 동작하기 때문에, 의도한 대로 `Future<String>`을 반환 값으로 가져올 수 있다

### 참고
- https://www.baeldung.com/java-completablefuture


